### PR TITLE
Search as you type

### PR DESCRIPTION
**Pull request by sharabash**

### CODE REVIEW

File: app/controllers/suggest_controller.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# This class is an endpoint for serving ayah suggestions from search text
+class SuggestController < ApplicationController
```
Comment (mmahalwy) [inline app/controllers/suggest_controller.rb]: I'd suggest you have an action called `suggest` in the `SearchController`

File: app/controllers/suggest_controller.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    lang = params[:l] || params[:lang] || 'en' # TODO we need to explicitly pass in lang, this default shouldn't be around forever
+    query = params[:q] || params[:query]
+    size = ( params[:s] || params[:size] || '5' ).to_i
+
+    # clean the query text
+    query = Search::Query::Query.new( query )
+    qtext = query[:query]
+
+
```
Comment (mmahalwy) [inline app/controllers/suggest_controller.rb]: :scissors:

File: app/controllers/suggest_controller.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+class SuggestController < ApplicationController
+  def query
+
+    lang = params[:l] || params[:lang] || 'en' # TODO we need to explicitly pass in lang, this default shouldn't be around forever
+    query = params[:q] || params[:query]
+    size = ( params[:s] || params[:size] || '5' ).to_i
+
+    # clean the query text
+    query = Search::Query::Query.new( query )
+    qtext = query[:query]
```
Comment (mmahalwy) [inline app/controllers/suggest_controller.rb]: variable name should be more explicit, like `query_text`

File: app/controllers/suggest_controller.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+
+    if qtext.length < 3
+      return render json: {error: 'not enough characters'}, status: 400
+    end
+
+    if Rails.env.production?
```
Comment (mmahalwy) [inline app/controllers/suggest_controller.rb]: This shouldn't be here. See what I did here: https://github.com/quran/quran.com-api/blob/dev/app/models/search.rb#L1

File: app/controllers/suggest_controller.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          #resource: hit['_source']['resource_id'],
+          text: text,
+          #full: hit['_source']['text'],
+          #hlit: hit['highlight']['text.autocomplete'],
+          href: href,
+          ayah: ayah
+        }
+        processed.push( h )
+      end
+    end
```
Comment (mmahalwy) [inline app/controllers/suggest_controller.rb]: everything above from here should be abstracted into it's own builder model, like the ones here: https://github.com/quran/quran.com-api/tree/dev/app/models/search

File: app/models/concerns/searchable.rb [hunk 6]
```diff
@@ -2,8 +2,13 @@
 module Searchable
     extend ActiveSupport::Concern
 
-    @@models = [ 'Content::Translation', 'Content::Transliteration', 'Content::TafsirAyah',
-                 'Quran::TextFont' ]
+    @@models = [
+        'Content::Translation',
```
Comment (mmahalwy) [inline app/models/concerns/searchable.rb]: your indentation is off

File: config/routes.rb [hunk 7]
```diff
@@ -6,6 +6,7 @@
   end
 
   get 'search', to: 'search#query'
+  get 'suggest', to: 'suggest#query'
```
Comment (mmahalwy) [inline config/routes.rb]: should have this in the search controller

File: app/controllers/suggest_controller.rb (hunk 5 above)
Comment (sharabash) [inline app/controllers/suggest_controller.rb]: thats a lot of abstraction for just 100 lines of code O_O i'm looking at it but i don't know exactly what to do; this isn't nearly as convoluted as my first attempt at search was; it's just a straightforward request with the ES client and a straightforward mapping of the results to a simple json object. i don't know where to start so I'll get back to this after looing at redux

Comment (mmahalwy) [inline app/controllers/suggest_controller.rb]: Still, I think we leave 100 lines here, and another there, then we have code that no one can read and understand. I am happy to help on this PR if needed. If I did not attempt to rewrite search, I would not know what's going on here - so abstraction is not for making things clean - it's for organization of understanding. Knowing why this is there and what it does.

**Comment by sharabash** [comment]
ok @mmahalwy i put the suggest logic in a model -- is this okay yet? :-(

**Comment by mmahalwy** [comment]
mashallah! This is awesome :) :+1: :+1: :+1: :+1: :+1: :+1:

**Pull request merged**